---
title: "Tutorial 3: scATAC Longitudinal Data (n=4 and 6 weeks follow-up)"
output:
  pdf_document:
    toc: yes
    number_sections: yes
  html_document:
    toc: yes
    df_print: paged
  word_document:
    toc: yes
urlcolor: blue
vignette: |
  %\VignetteIndexEntry{ReferenceManual}
  %\VignetteEngine{knitr::rmarkdown} %\VignetteEncoding{UTF-8}
editor_options: 
  markdown: 
    wrap: sentence
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# <a name="example3"></a> scATAC Longitudinal Data (n=4 and 6 weeks follow-up)

This tutorial allows to explore single cell ATACseq genscore data measured from 4 healthy donors over 6 timepoints (week 2-7).
Single cell ATAC data available at [GSE190992\*](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE190992).
(1) AIFI-scATAC-PBMC-FinalData.Rda (2) [AIFI-Metadata.Rda](https://github.com/aifimmunology/PALMO/blob/data/data/AIFI-Metadata.Rda) (clinical metadata).
Longitudinal dataset have 4 donors and 18 samples.
To infer the variations at single cell ATAC please follow following steps.
*scATAC genescore for longitudinal data was retrieved using ArchR (<https://www.archrproject.com/>) package. The genescore matrix was used as input for PALMO to explore longitudinal stability or variations across celltypes.*

*(Note:If any issue with data access or needed RAW files please [Contact Us](mailto:suhas.vasaikar@alleninstitute.org))*.

## Load Library

    library("PALMO")

## Load scATAC genescore data and assign paramaters (Time \<30sec)

Load genescorematrix (pseudo-bulk) from ArchR or relevant tools (Aggregated peaks at gene level).
We used ArchR to generate genescore matrix from scATAC peakmatrix of longitudinal data by each sample and celltype level.

    #scATAC object
    load("data/AIFI-scATAC-PBMC-FinalData.Rda")
    #log-normalized
    datamatrix <- log2(scatac_gm+1)

    #Load annotation data
    load("data/AIFI-Metadata.Rda")

## Create PALMO object (Time \<30sec)

First create the PALMO S4 object using input scATAC genescore data and annotation data frame.
The expression data frame columns merged with input annotation data frame.
Only overlapping samples kept.

    #Create PALMO object
    palmo_obj <- createPALMOobject(anndata=ann, data=datamatrix)

Assign sample_column, donor_column and time_column variables in this step.

    #Assign Sample, PTID and Time parameters
    palmo_obj <- annotateMetadata(data_object=palmo_obj, 
                  sample_column= "Sample", donor_column= "PTID",
                  time_column= "Time")

For aggregated genescore data merge data column with `Sample` annotation mentioned in annotation dataframe.

    #Sample overlap and final matrix
    palmo_obj <- mergePALMOdata(data_object=palmo_obj, datatype="singlecell")

## CV profile (Time \~2min)

To calculate longitudinal stability within donor, first visualize the CV vs mean distribution for given data.
Define CV threshold using the histogram or selecting few features of interest CV as cut-off.
Here we used housekeeping genes GAPDH and ACTB to define mean and CV threshold.

    palmo_obj <- cvCalcSCProfile(data_object=palmo_obj,
                                 housekeeping_genes=c("GAPDH", "ACTB"),
                                 fileName="scatac")

<br> ![img](imgs/Tutorial-3-cvDistribution-1.png) <br>

    palmo_obj <- cvCalcSCProfile(data_object=palmo_obj,
                                 housekeeping_genes=c("GAPDH", "ACTB"),
                                 meanThreshold = 0.1,
                                 fileName="scatac")
                                 

<br> ![img](imgs/Tutorial-3-cvDistribution-2.png) <br>

## **Optional step**.
To check the donor or participant wise CV profile run the CV sample profile analysis.
Plots are saved in output directory if directory not mentioned.

    #Sample Celltype Mean-CV plot
    cvSCsampleprofile(data_object=palmo_obj, meanThreshold = 0.1,
                      cvThreshold = 10, fileName="scatac")

## Features contributing towards donor variations (Time \~ 8min)

Variance decomposition analysis was performed to identify the features associated with attributes of interest such as PTID (participants) , Time (weeks), and group (celltype).
The variance explained by each gene towards the `featureSet` of interest given in percentage.
*(Note: To reduce the processing time use nodes cl=8 (or greater) and lmer_control=TRUE)*

    #Variance decomposition
    featureSet <- c("PTID","Time","group")
    palmo_obj <- lmeVariance(data_object=palmo_obj,
                             featureSet=featureSet,
                             meanThreshold=0.1, cl=4,
                             fileName="scatac")
    var_decomp <- palmo_obj@result$variance_decomposition
    head(var_decomp[,featureSet])
    #                 PTID      Time     group
    #FIRRE        88.88239 0.2606612 2.8339971
    #XIST         87.05064 0.3986387 0.5178851
    #RHD          81.77230 0.3641998 1.4574140
    #GSTM1        80.29766 0.1786628 6.2277645
    #ZNF705D      70.69163 0.4342325 0.4752960
    #LOC105376805 66.62977 1.2741294 9.8918653

<br> ![img](imgs/Tutorial-3-variancePlot.png){width="50%" height="50%"} <br>

The top 15 features contributing to donor, time or group/celltype attributes variance can be visualized in barplot.

    #Variance contributing features
    plots <- variancefeaturePlot(vardata=var_decomp, featureSet=featureSet,
                                 cols=c("purple", "darkgreen", "cyan"),
                                 ncol=3)

<br> ![img](imgs/Tutorial-3-Donor-variancePlot.png) <br>

Gene expression plot for interested features can be visualized by donors or by time.

    #Top genes
    plots <- gene_featureplot(data_object=palmo_obj,
                              featureList=c("FIRRE", "XIST", 
                                            "DPP6", "THSD7A",
                                            "SPIB", "KLF4"),
                              x_group_by="PTID", var_oi="Time",
                              x_text_angle=90)

<br> ![img](imgs/Tutorial-3a-geneplot.png) <br>

Gene expression plot for interested features can be visualized by group (celltypes).

    gene_featureplot(data_object=palmo_obj, featureList="SPIB", facet_by="group")

<br> ![img](imgs/Tutorial-3b-geneplot.png) <br>

## Intra-donor variations over time (Time \~ 5min)

To calculate longitudinal stability within donor define meanThreshold and cvThreshold parameters as discussed above.
The analysis will calculate the CV across average group ("celltype") mentioned by user.
It will create a CV-Mean plots for individual donor over longitudinal timepoints that shows the highly variable and stable features in each donor.
The plots are stored in output directory if not mentioned.
It also provides a housekeeping genes position in CV-Mean plot across each celltype/group of interest.
*(Note: To reduce the processing time use nodes cl=8 (or greater))*

    palmo_obj <- cvCalcSC(data_object=palmo_obj,
                       meanThreshold=0.1, cvThreshold=10,
                       housekeeping_genes=c("GAPDH", "ACTB"),
                       fileName="scatac")
                       

<br> ![img](imgs/Tutorial-3-CV-distribution.png) <br>

*(Note: Black dots, all genes; red dots, genes with given mean and CV threshold; blue dots, housekeeping genes)*

## Find stable and variable features in longitudinal data (Time 30sec)

To identify Stable and variable features across celltype or group of interest CV values for each donor at celltype retrieved from above step.
Minimum number of donors (donorThreshold) and group threshold (groupThreshold) defined to retrieve features with consensus stable/variable profile.
Here, donorThreshold=4 and groupThreshold=28 (4\*14/2) is selected.

    #Find stable and variable features in longitudinal data
    palmo_obj <- StableFeatures(data_object=palmo_obj,
                                  cvThreshold=10,
                                  donorThreshold=4,
                                  topFeatures=25,
                                  fileName="scatac")
    stable_genes <- palmo_obj@result$stable_genes

<br> ![img](imgs/Tutorial-3-Stable-Plot.png) <br>

    palmo_obj <- VarFeatures(data_object=palmo_obj,
                            cvThreshold=10,
                            donorThreshold=4, groupThreshold=28, 
                            topFeatures=25,
                            fileName="scatac")
    var_genes <- palmo_obj@result$var_genes

<br> ![img](imgs/Tutorial-3-Variable-Plot.png) <br>

To visualize how these stable or variable features performing, users can reduce high dimensional single cell ATAC data into lower dimensions using selected Top features from stable/variable analysis.
The UMAP plot shows whether identified stable features can reproduce celltype/group identity based on few but stable features.

## Circos CV plot (Time \~ 10sec)

The top features or gene families can be visualized in circos plot for easy comparison.

    geneList <- c("HLA-A","HLA-B","HLA-C","HLA-DRA","HLA-DPA1","HLA-DRB1",
                  "ACTB","GAPDH")
    plotmatrix <- genecircosPlot(data_object=palmo_obj, geneList=geneList, 
                                 colorThreshold=15)

To order the circos plot by user-defined group order follow following steps.

    #order by user-defined group order 
    celltype_oi <- c("CD4_Naive","CD4_TEM","CD4_TCM","CD4_CTL",
                  "CD8_Naive","CD8_TEM","CD8_TCM","Treg","MAIT","gdT",
                  "NK", "NK_CD56bright",
                  "B_naive","B_memory", "B_intermediate",
                  "CD14_Mono","CD16_Mono",
                  "cDC2","pDC")
    plotmatrix <- genecircosPlot(data_object=palmo_obj, geneList=geneList, 
                                 group_oi=celltype_oi, colorThreshold=15)

<br> ![img](imgs/Tutorial-3-celltype-circularPlot.png) <br>

# <a name="session"></a> Session Info

    sessionInfo()
    #> R version 4.1.2 (2021-11-01)
    #> Platform: x86_64-apple-darwin17.0 (64-bit)
    #> Running under: macOS Big Sur 10.16
    #> 
    #> Matrix products: default
    #> BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib
    #> LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib
    #> 
    #> locale:
    #> [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
    #> 
    #> attached base packages:
    #> [1] grid      stats     graphics  grDevices utils     datasets  methods  
    #> [8] base     
    #> 
    #> other attached packages:
    #> [1] PALMO_0.1.1
    #> 
    #> loaded via a namespace (and not attached):
    #>   [1] utf8_1.2.2                  reticulate_1.25            
    #>   [3] tidyselect_1.1.2            lme4_1.1-31                
    #>   [5] htmlwidgets_1.5.4           Rtsne_0.16                 
    #>   [7] munsell_0.5.0               codetools_0.2-18           
    #>   [9] ragg_1.2.2                  ica_1.0-2                  
    #>  [11] future_1.26.1               miniUI_0.1.1.1             
    #>  [13] withr_2.5.0                 spatstat.random_2.2-0      
    #>  [15] colorspace_2.0-3            progressr_0.10.1           
    #>  [17] Biobase_2.54.0              knitr_1.40                 
    #>  [19] rstudioapi_0.14             SingleCellExperiment_1.16.0
    #>  [21] Seurat_4.1.1                stats4_4.1.2               
    #>  [23] ROCR_1.0-11                 tensor_1.5                 
    #>  [25] listenv_0.8.0               MatrixGenerics_1.6.0       
    #>  [27] GenomeInfoDbData_1.2.7      polyclip_1.10-0            
    #>  [29] farver_2.1.1                pheatmap_1.0.12            
    #>  [31] rprojroot_2.0.3             parallelly_1.32.0          
    #>  [33] vctrs_0.5.1                 generics_0.1.3             
    #>  [35] xfun_0.33                   GenomeInfoDb_1.30.1        
    #>  [37] R6_2.5.1                    doParallel_1.0.17          
    #>  [39] clue_0.3-61                 DelayedArray_0.20.0        
    #>  [41] bitops_1.0-7                spatstat.utils_2.3-1       
    #>  [43] cachem_1.0.6                assertthat_0.2.1           
    #>  [45] promises_1.2.0.1            scales_1.2.1               
    #>  [47] rgeos_0.5-9                 gtable_0.3.1               
    #>  [49] globals_0.15.0              goftest_1.2-3              
    #>  [51] rlang_1.0.6                 systemfonts_1.0.4          
    #>  [53] GlobalOptions_0.1.2         splines_4.1.2              
    #>  [55] lazyeval_0.2.2              spatstat.geom_2.4-0        
    #>  [57] broom_0.8.0                 yaml_2.3.5                 
    #>  [59] reshape2_1.4.4              abind_1.4-5                
    #>  [61] modelr_0.1.8                backports_1.4.1            
    #>  [63] httpuv_1.6.5                tools_4.1.2                
    #>  [65] ggplot2_3.4.0               ellipsis_0.3.2             
    #>  [67] spatstat.core_2.4-4         jquerylib_0.1.4            
    #>  [69] RColorBrewer_1.1-3          BiocGenerics_0.40.0        
    #>  [71] ggridges_0.5.3              Rcpp_1.0.9                 
    #>  [73] plyr_1.8.7                  zlibbioc_1.40.0            
    #>  [75] RCurl_1.98-1.8              purrr_0.3.4                
    #>  [77] rpart_4.1.16                deldir_1.0-6               
    #>  [79] pbapply_1.5-0               GetoptLong_1.0.5           
    #>  [81] cowplot_1.1.1               S4Vectors_0.32.4           
    #>  [83] zoo_1.8-10                  SummarizedExperiment_1.24.0
    #>  [85] SeuratObject_4.1.0          haven_2.5.0                
    #>  [87] ggrepel_0.9.1               cluster_2.1.3              
    #>  [89] factoextra_1.0.7            fs_1.5.2                   
    #>  [91] magrittr_2.0.3              data.table_1.14.2          
    #>  [93] scattermore_0.8             circlize_0.4.15            
    #>  [95] lmtest_0.9-40               reprex_2.0.1               
    #>  [97] RANN_2.6.1                  fitdistrplus_1.1-8         
    #>  [99] matrixStats_0.62.0          hms_1.1.2                  
    #> [101] patchwork_1.1.1             mime_0.12                  
    #> [103] evaluate_0.16               xtable_1.8-4               
    #> [105] readxl_1.4.0                IRanges_2.28.0             
    #> [107] gridExtra_2.3               shape_1.4.6                
    #> [109] compiler_4.1.2              tibble_3.1.8               
    #> [111] KernSmooth_2.23-20          crayon_1.5.1               
    #> [113] minqa_1.2.4                 htmltools_0.5.3            
    #> [115] mgcv_1.8-40                 later_1.3.0                
    #> [117] tzdb_0.3.0                  tidyr_1.2.1                
    #> [119] lubridate_1.8.0             DBI_1.1.3                  
    #> [121] tweenr_1.0.2                dbplyr_2.2.1               
    #> [123] ComplexHeatmap_2.10.0       MASS_7.3-57                
    #> [125] MAST_1.20.0                 boot_1.3-28                
    #> [127] Matrix_1.5-1                readr_2.1.2                
    #> [129] cli_3.4.0                   parallel_4.1.2             
    #> [131] igraph_1.3.2                GenomicRanges_1.46.1       
    #> [133] forcats_0.5.2               pkgconfig_2.0.3            
    #> [135] pkgdown_2.0.7               sp_1.5-0                   
    #> [137] plotly_4.10.0               spatstat.sparse_2.1-1      
    #> [139] xml2_1.3.3                  foreach_1.5.2              
    #> [141] bslib_0.3.1                 XVector_0.34.0             
    #> [143] rvest_1.0.2                 stringr_1.4.1              
    #> [145] digest_0.6.29               sctransform_0.3.3          
    #> [147] RcppAnnoy_0.0.19            spatstat.data_2.2-0        
    #> [149] rmarkdown_2.14              cellranger_1.1.0           
    #> [151] leiden_0.4.2                uwot_0.1.14                
    #> [153] shiny_1.7.1                 rjson_0.2.21               
    #> [155] nloptr_2.0.3                lifecycle_1.0.3            
    #> [157] nlme_3.1-158                jsonlite_1.8.0             
    #> [159] desc_1.4.1                  viridisLite_0.4.1          
    #> [161] fansi_1.0.3                 pillar_1.8.1               
    #> [163] lattice_0.20-45             fastmap_1.1.0              
    #> [165] httr_1.4.4                  survival_3.3-1             
    #> [167] glue_1.6.2                  png_0.1-7                  
    #> [169] iterators_1.0.14            ggforce_0.3.3              
    #> [171] stringi_1.7.8               sass_0.4.1                 
    #> [173] textshaping_0.3.6           memoise_2.0.1              
    #> [175] dplyr_1.0.10                tidyverse_1.3.1            
    #> [177] irlba_2.3.5                 future.apply_1.9.0
*The analysis time reported in tutorials were based on macOS with 2.3GHz 8-core Intel Core i9 processor and 32GB RAM memory.*

\newpage
\newpage